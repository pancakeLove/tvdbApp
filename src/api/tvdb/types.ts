import {AxiosError} from 'axios';

export interface LoginResponse {
  readonly token: string;
  readonly error?: AxiosError;
}

export interface SearchResultItem {
  readonly aliases: string;
  readonly banner: string;
  readonly firstAired: string;
  readonly id: number;
  readonly image: string;
  readonly network: string;
  readonly overview: string;
  readonly poster: string;
  readonly seriesName: string;
  readonly slug: string;
  readonly status: string;
}
export interface SearchResponse {
  readonly data: SearchResultItem[];
  readonly error?: AxiosError;
}

export interface SeriesDetails {
  readonly added: string;
  readonly airsDayOfWeek: string;
  readonly airsTime: string;
  readonly aliases: [string];
  readonly banner: string;
  readonly firstAired: string;
  readonly id: number;
  readonly imdbId: string;
  readonly network: string;
  readonly networkId: string;
  readonly overview: string;
  readonly rating: string;
  readonly runtime: string;
  readonly seriesId: string;
  readonly seriesName: string;
  readonly siteRating: number;
  readonly siteRatingCount: number;
  readonly slug: string;
  readonly status: string;
  readonly zap2itId: string;
}

export interface SeriesDetailsResponse {
  readonly data: SeriesDetails;
  readonly error?: AxiosError;
}
