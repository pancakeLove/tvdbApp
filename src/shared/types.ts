export interface GeneralError {
  readonly error: boolean;
  readonly message: string;
}
