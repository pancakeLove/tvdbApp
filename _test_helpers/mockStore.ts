import configureMockStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';

export const mockStore = (state: Object, tvdbApi: any, saga: any) => {
  const context = {
    tvdbApi,
  };
  const sagaMiddleware = createSagaMiddleware();
  sagaMiddleware.setContext(context);
  const mockStore = configureMockStore([sagaMiddleware]);
  const store = mockStore(state);
  sagaMiddleware.run(saga);
  return {...store};
};
