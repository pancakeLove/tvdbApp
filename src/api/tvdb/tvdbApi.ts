import {AxiosRequestConfig} from 'axios';
// @ts-ignore
import {APIKey, Userkey, Username} from '@env';
import {LoginResponse, SearchResponse, SeriesDetailsResponse} from './types';
import {sendRequest} from '../apiHelpers';

const requestConfig: Partial<AxiosRequestConfig> = {
  baseURL: 'https://api.thetvdb.com/',
  timeout: 10000,
  method: 'get',
};

export const tvdbApi = () => ({
  fetchToken: async () => {
    let config: AxiosRequestConfig = {
      ...requestConfig,
      method: 'post',
      url: 'login',
      data: {
        apikey: APIKey,
        userkey: Userkey,
        username: Username,
      },
    };
    return (await sendRequest(config)) as Promise<LoginResponse>;
  },
  search: async (query: string, jwt: string) => {
    let config: AxiosRequestConfig = {
      ...requestConfig,
      url: 'search/series',
      params: {name: query},
      headers: {Authorization: `Bearer ${jwt}`},
    };
    return (await sendRequest(config)) as SearchResponse;
  },
  details: async (id: number, jwt: string) => {
    let config: AxiosRequestConfig = {
      ...requestConfig,
      url: `series/${id}`,
      params: {id},
      headers: {Authorization: `Bearer ${jwt}`},
    };
    return (await sendRequest(config)) as SeriesDetailsResponse;
  },
});
