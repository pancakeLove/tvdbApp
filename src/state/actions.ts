import {createStandardAction} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export enum TvDbActionTypes {
  LOGIN_REQUEST = 'LOGIN_REQUEST',
  LOGIN_SUCCESS = 'LOGIN_SUCCESS',
  LOGIN_FAILURE = 'LOGIN_FAILURE',
  SEARCH_REQUEST = 'SEARCH_REQUEST',
  SEARCH_SUCCESS = 'SEARCH_SUCCESS',
  DETAILS_REQUEST = 'DETAILS_REQUEST',
  DETAILS_SUCCESS = 'DETAILS_SUCCESS',
  API_ERROR = 'API_ERROR',
  IS_LOADING = 'IS_LOADING',
}

export const tvDbActions = {
  loginRequest: createStandardAction(TvDbActionTypes.LOGIN_REQUEST)<null>(),
  loginSuccess: createStandardAction(TvDbActionTypes.LOGIN_SUCCESS)<string>(),
  loginFailure: createStandardAction(TvDbActionTypes.LOGIN_FAILURE)(),
  searchRequest: createStandardAction(TvDbActionTypes.SEARCH_REQUEST)<string>(),
  searchSuccess: createStandardAction(TvDbActionTypes.SEARCH_SUCCESS)<
    SearchResultItem[]
  >(),
  detailsRequest: createStandardAction(
    TvDbActionTypes.DETAILS_REQUEST,
  )<number>(),
  detailsSuccess: createStandardAction(
    TvDbActionTypes.DETAILS_SUCCESS,
  )<SeriesDetails | null>(),
  apiError: createStandardAction(
    TvDbActionTypes.API_ERROR,
  )<GeneralError | null>(),
  isLoading: createStandardAction(TvDbActionTypes.IS_LOADING)<boolean>(),
};
