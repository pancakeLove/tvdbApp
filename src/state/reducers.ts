import {Reducer} from 'redux';
import {TvDbActionTypes} from './actions';
import {Action} from 'typesafe-actions';
import {SearchResultItem, SeriesDetails} from '../api/tvdb/types';
import {GeneralError} from '../shared/types';

export interface TvDbState {
  loginInit: boolean;
  token: string;
  loginError: boolean;
  searchResult?: SearchResultItem[];
  apiError?: GeneralError;
  isLoading: boolean;
  seriesDetails?: SeriesDetails;
}

const initialState = {
  loginInit: false,
  token: '',
  loginError: false,
  searchResult: [],
  apiError: {error: false, message: ''},
  isLoading: false,
};

export interface DispatchAction extends Action<TvDbActionTypes> {
  payload: Partial<TvDbState>;
}

export const tvDbReducer: Reducer<TvDbState, DispatchAction> = (
  state = initialState,
  action,
) => {
  switch (action?.type) {
    case TvDbActionTypes.LOGIN_REQUEST:
      return {...state, loginError: false, loginInit: true};
    case TvDbActionTypes.LOGIN_SUCCESS:
      return {...state, token: action.payload as string, isLoading: false};
    case TvDbActionTypes.LOGIN_FAILURE:
      return {
        ...state,
        token: '',
        loginError: true,
      };
    case TvDbActionTypes.SEARCH_SUCCESS:
      return {
        ...state,
        searchResult: action.payload as SearchResultItem[],
        isLoading: false,
      };
    case TvDbActionTypes.DETAILS_SUCCESS:
      return {
        ...state,
        seriesDetails: action.payload as SeriesDetails,
        isLoading: false,
      };
    case TvDbActionTypes.API_ERROR:
      return {
        ...state,
        apiError: action.payload as GeneralError,
        isLoading: false,
      };
    default:
      return state;
  }
};
