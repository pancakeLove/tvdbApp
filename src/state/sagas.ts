import {
  takeLatest,
  call,
  put,
  all,
  select,
  getContext,
} from 'redux-saga/effects';
import {tvDbActions, TvDbActionTypes} from './actions';
import {
  LoginResponse,
  SearchResponse,
  SeriesDetailsResponse,
} from '../api/tvdb/types';
import {AxiosError} from 'axios';
import {TvDbState} from './reducers';
import {PayloadAction} from 'typesafe-actions';

function* handleApiError(error: AxiosError<any>) {
  let message: string = '';
  if (error.response) {
    message = 'Server responded with non 2xx status: ' + error.response.status;
  } else if (error.request) {
    message = 'Timeout received, possible slow network';
  } else {
    message = error.message;
  }
  yield put(tvDbActions.apiError({error: true, message}));
}

export function* loginWatcher() {
  yield takeLatest(TvDbActionTypes.LOGIN_REQUEST, loginWorker);
}
export function* loginWorker() {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  //console.log('tvdbApi from context:', tvdbApi);
  const response: LoginResponse = yield call(tvdbApi().fetchToken);
  if (response.error) {
    yield put({type: TvDbActionTypes.LOGIN_FAILURE});
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.loginSuccess(response.token));
  }
}

export function* searchWatcher() {
  yield takeLatest(TvDbActionTypes.SEARCH_REQUEST, searchWorker);
}
function* searchWorker(action: PayloadAction<any, string>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SearchResponse = yield call(
    tvdbApi().search,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    /* todo: should be some limitation to number of results, eg:
     yield put(tvDbActions.searchSuccess(response.data?.slice(0, 100)));
     */
    yield put(tvDbActions.searchSuccess(response.data));
  }
}

export function* fetchDetailsWatcher() {
  yield takeLatest(TvDbActionTypes.DETAILS_REQUEST, fetchDetailsWorker);
}
function* fetchDetailsWorker(action: PayloadAction<any, number>) {
  yield put(tvDbActions.isLoading(true));
  const tvdbApi = yield getContext('tvdbApi');
  const token = yield select((state: TvDbState) => state.token);
  const response: SeriesDetailsResponse = yield call(
    tvdbApi().details,
    action.payload,
    token,
  );
  if (response.error) {
    yield handleApiError(response.error);
    yield put(tvDbActions.isLoading(false));
  } else {
    yield put(tvDbActions.detailsSuccess(response.data));
  }
}

export function* tvDbSagas() {
  yield all([loginWatcher(), searchWatcher(), fetchDetailsWatcher()]);
}
