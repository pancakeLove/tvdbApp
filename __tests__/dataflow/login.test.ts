import {tvDbActions} from '../../src/state/actions';
import {tvDbSagas} from '../../src/state/sagas';
import {mockStore} from '../../_test_helpers/mockStore';
import {LoginResponse} from '../../src/api/tvdb/types';
import {AxiosError} from 'axios';
import {tvDbReducer} from '../../src/state/reducers';

describe('dataflow for login', () => {
  it('should dispatch a successful request to login', async () => {
    const loginSuccessResponse: LoginResponse = {
      token: 'token',
      error: undefined,
    };
    const tvdbApi = () => ({
      fetchToken: jest.fn().mockReturnValue(loginSuccessResponse),
    });
    const store = mockStore(tvDbReducer, tvdbApi, tvDbSagas);
    await store.dispatch(tvDbActions.loginRequest(null));
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should dispatch a failing request to login', async () => {
    const loginFailingResponse: LoginResponse = {
      token: '',
      error: {} as AxiosError,
    };
    const tvdbApi = () => ({
      fetchToken: jest.fn().mockReturnValue(loginFailingResponse),
    });
    const store = mockStore({}, tvdbApi, tvDbSagas);
    await store.dispatch(tvDbActions.loginRequest(null));
    expect(store.getActions()).toMatchSnapshot();
  });
});
