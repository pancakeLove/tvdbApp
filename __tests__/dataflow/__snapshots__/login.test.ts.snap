// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataflow for login should dispatch a failing request to login 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": null,
    "type": "LOGIN_REQUEST",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "type": "LOGIN_FAILURE",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "error": true,
      "message": undefined,
    },
    "type": "API_ERROR",
  },
  Object {
    "meta": undefined,
    "payload": false,
    "type": "IS_LOADING",
  },
]
`;

exports[`dataflow for login should dispatch a successful request to login 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": null,
    "type": "LOGIN_REQUEST",
  },
  Object {
    "meta": undefined,
    "payload": true,
    "type": "IS_LOADING",
  },
  Object {
    "meta": undefined,
    "payload": "token",
    "type": "LOGIN_SUCCESS",
  },
]
`;
